src/linear_algebra/symplectic_group.lean

```diff
@@ -0,0 +1,252 @@
+`symplectic_group`: the group of symplectic matrices
+
+## Implementation Notes
+
+## TODO
+* Every symplectic matrix has determinant 1.
+* For `n = 1` the symplectic group coincides with the special linear group.
+-/
+
+open_locale matrix
+
+variables {l R : Type*}
+
+lemma pm_one_unit [ring R] {x : R} (h : x = 1 ∨ x = -1) : is_unit x :=
+begin
+  cases h,
+  { simp only [h, is_unit_one] },
+  { use -1,
+    simp only [h, units.coe_neg_one] }
+end
+
+section J_matrix_lemmas
+
+namespace matrix
+
+variables (l) [decidable_eq l]
+
+variables (R) [comm_ring R]
+
+/-- The matrix defining the canonical skew-symmetric bilinear form. -/
+def J : matrix (l ⊕ l) (l ⊕ l) R := matrix.from_blocks 0 (-1) 1 0
+
+@[simp] lemma J_transpose : (J l R)ᵀ = - (J l R) :=
+begin
+  unfold J,
+  rw [from_blocks_transpose, ←neg_one_smul R (from_blocks _ _ _ _), from_blocks_smul,
+    matrix.transpose_zero, matrix.transpose_one, transpose_neg],
+  simp [from_blocks],
+end
+
+variables [fintype l]
+
+lemma J_squared : (J l R) ⬝ (J l R) = -1 :=
+begin
+  unfold J,
+  rw from_blocks_multiply,
+  simp only [matrix.zero_mul, matrix.neg_mul, zero_add, neg_zero', matrix.one_mul, add_zero],
+  rw [← neg_zero, ← matrix.from_blocks_neg,← from_blocks_one],
+end
+
+lemma J_inv : (J l R)⁻¹ = -(J l R) :=
+begin
+  refine matrix.inv_eq_right_inv _,
+  rw [matrix.mul_neg, J_squared],
+  exact neg_neg 1,
+end
+
+variables (R)
+variables [no_zero_divisors R] [nontrivial R]
+
+lemma J_det [rc : fact (ring_char R ≠ 2)] : det (J l R) = 1 ∨ det (J l R) = - 1:=
+begin
+  have H : (det (J l R)) * (det (J l R)) = 1 := by
+  { rw [←det_mul, J_squared],
+    rw [←one_smul R (-1 : matrix _ _ R)],
+    rw [smul_neg, ←neg_smul, det_smul],
+    simp only [fintype.card_sum, det_one, mul_one],
+    rw neg_one_pow_eq_one_iff_even (ring.neg_one_ne_one_of_char_ne_two (fact.elim rc)),
+    exact even_add_self _ },
+  rw [←sq_eq_one_iff, pow_two],
+  exact H,
+end
+
+lemma J_det_unit [fact (ring_char R ≠ 2)] : is_unit (det (J l R)) :=
+  pm_one_unit (J_det l R)
+
+end matrix
+
+end J_matrix_lemmas
+
+open matrix
+
+variables (l) (R) [decidable_eq l] [fintype l] [comm_ring R]
+
+/-- The group of symplectic matrices over a ring `R`. -/
+def symplectic_group : submonoid (matrix (l ⊕ l) (l ⊕ l)  R) :=
+{ carrier := { A | A ⬝ (J l R) ⬝ Aᵀ = J l R},
+  mul_mem' :=
+  begin
+    intros a b ha hb,
+    simp only [mul_eq_mul, set.mem_set_of_eq, transpose_mul] at *,
+    rw [←matrix.mul_assoc, a.mul_assoc, a.mul_assoc, hb],
+    exact ha,
+  end,
+  one_mem' := by simp }
+
+variables {l} {R}
+
+namespace symplectic_group
+
+lemma mem_iff {A : matrix (l ⊕ l) (l ⊕ l)  R} :
+  A ∈ symplectic_group l R ↔ A ⬝ (J l R) ⬝ Aᵀ = J l R :=
+by simp [symplectic_group]
+
+instance coe_matrix : has_coe (symplectic_group l R) (matrix (l ⊕ l) (l ⊕ l)  R) := ⟨subtype.val⟩
+
+instance coe_fun : has_coe_to_fun (symplectic_group l R) (λ _, (l ⊕ l) → (l ⊕ l) → R) :=
+{ coe := λ A, A.val }
```

Comment:
I don't think this instance is a good idea

Reply:
I was copying the coercions from the unitary group, but after writing the rest I tend to agree: it makes it impossible to use `ext` and probably other things as well because `ext` sends you to proving `A i j = B i j` and not `A = B` for matrices `A, B`.

Reply:
The ones about the unitary group should probably be deleted too, but that's out of scope for this PR.

Reply:
Deleted the instance.